test/e2e/framework/job/wait.go

```diff
@@ -60,19 +59,38 @@ func WaitForJobPodsSucceeded(ctx context.Context, c clientset.Interface, ns, job
 
 // waitForJobPodsInPhase wait for all pods for the Job named JobName in namespace ns to be in a given phase.
 func waitForJobPodsInPhase(ctx context.Context, c clientset.Interface, ns, jobName string, expectedCount int32, phase v1.PodPhase, timeout time.Duration) error {
-	return wait.PollUntilContextTimeout(ctx, framework.Poll, timeout, false, func(ctx context.Context) (bool, error) {
-		pods, err := GetJobPods(ctx, c, ns, jobName)
-		if err != nil {
-			return false, err
-		}
-		count := int32(0)
-		for _, p := range pods.Items {
-			if p.Status.Phase == phase {
-				count++
+	return framework.Gomega().
+		Eventually(ctx, func(ctx context.Context) (*v1.PodList, error) {
+			return GetJobPods(ctx, c, ns, jobName)
+		}).
+		WithPolling(framework.Poll).
+		WithTimeout(timeout).
+		Should(framework.MakeMatcher(func(pods *v1.PodList) (func() string, error) {
+			var nonMatchingPods, matchingPods []v1.Pod
+			for _, p := range pods.Items {
+				if p.Status.Phase == phase {
+					matchingPods = append(matchingPods, p)
+				} else {
+					nonMatchingPods = append(nonMatchingPods, p)
+				}
 			}
-		}
-		return count == expectedCount, nil
-	})
+
+			if len(matchingPods) > int(expectedCount) {
+				return func() string {
+					return fmt.Sprintf("expected at most %d pods in phase %s, %d out of %d were not:\n%s",
+						expectedCount, phase, len(matchingPods), len(pods.Items), format.Object(matchingPods, 1))
+				}, nil
+			}
+
+			if len(matchingPods) < int(expectedCount) {
+				return func() string {
+					return fmt.Sprintf("expected at least %d pods in phase %s, %d out of %d were not:\n%s",
+						expectedCount, phase, len(nonMatchingPods), len(pods.Items), format.Object(nonMatchingPods, 1))
+				}, nil
+			}
+
```

Comment:
Yeah, this adds complexity, but I understand the motivation - to help to understand the issues.

Maybe we could achieve the same debuggability with fewer `if` if we accummulate the count on map `podsByPhase[v1.PodPhase]int`, and then just check `if len(podsByPhase[phase]) == int(expectedCount)`, wdyt?

Reply:
> Yeah, this adds complexity, but I understand the motivation - to help to understand the issues.

Yes, it is my motivation.  @kaisoz @mimowo 
FYI: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/pod/wait.go#L352

Reply:
> returning the complete list of pods in case of failure. 
It only prints unexpected pods, not a complete list of pods.

It can help us filter out pods that do not meet the requirements and reduce our debugging work.

Reply:
> , wdyt? In case of an error we can just print the entire map.

I think it is unnecessary to dump the whole map. We may just need to know which pods are in the unexpected phase. It should be enough to solve flaky or failed tests.

Reply:
Yeah, but this requires more `ifs` and thus complicates the code, we need to balance that. In fact notice that in this PR we introduced dumping the entire object: https://github.com/kubernetes/kubernetes/pull/128079, and I think we can use this pattern more for Jobs - I remember debugging some of the e2e tests in the past and thinking "if only I knew the value of this field, it would be so much easier".

I think when something goes wrong you never exactly sure what might be useful. I can imagine the more granular information (how many failed, running, pending, succeeded) might be useful when debugging some scenarios.

Reply:
Got it. I will update it later.

Reply:
> I think when something goes wrong you never exactly sure what might be useful. I can imagine the more granular information (how many failed, running, pending, succeeded) might be useful when debugging some scenarios.

I agree, the more context the better

> Maybe we could achieve the same debuggability with fewer if if we accummulate the count on map podCountByPhase[v1.PodPhase]int, and then just check if len(podCountByPhase[phase]) == int(expectedCount), wdyt? In case of an error we can just print the entire map.

if the amount of pods in each phase is not going to be logged, why not reuse part of the original logic? (which is already quite simple):
```
count := int32(0)
for _, p := range pods.Items {
	if p.Status.Phase == phase {
		count++
	}
}
return count == expectedCount, nil
```

or do you mean that it eases debugging with a IDE?

Reply:
Your snippet lgtm. I was thinking about debugging based on logs, so was assuming we could log, I was thinking about something like:
```
podsCountPerPhase := make(map[v1.PodPhase]int)
for _, p := range pods.Items {
        podsCountPerPhase[p.Status.Phase]++
}
if len(podsCountPerPhase[phase]) == expectedCount {
   return true, nil
}
return false, fmt.Printf("unexpected number of pod counts: %v, expected: %v, map: %v", len(podsCountPerPhase[phase]), expectedCount, podsCountPerPhase)
```
This was the idea to balance the extra info with code simplicity, but I agree it is more code than in your snippet, so I'm perfectly ok either way

Reply:
> or do you mean that it eases debugging with a IDE?

Some flaky tests may be hard to reproduce in a local environment. In this case, we have to only depend on the log from the CI to debug the issue. A good log file should contain enough information to help us understand what happened during the test. All the e2e tests will delete all the resources after the test is done, so we can not know the final state of the resources before it is deleted. To be honest, the resource status is very important for us to understand the reason why the test failed. If we are very lucky, we can find the reason from the log file without debugging with a IDE.

This is why I want to dump all/a part of the pods. I'm unsure whether only having the number of pods is enough for us.

Reply:
> if the amount of pods in each phase is not going to be logged, why not reuse part of the original logic? (which is already quite simple):

I update my PR. I reuse the original logic but dump all pods if the match fails. Is it okay?

Reply:
sgtm